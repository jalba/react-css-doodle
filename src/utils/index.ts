export * from './test-utils';
export * from './template-utils';
