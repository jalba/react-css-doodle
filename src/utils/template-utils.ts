import { Func, Vars } from '../lib';

function isFunc(arg: Vars): arg is Func {
  return (arg as Func).call !== undefined;
}

export function valueToString(value: Vars): string {
  if (typeof value === 'number') {
    return value.toString();
  }

  if (typeof value === 'string') {
    return value;
  }

  if (isFunc(value)) {
    return valueToString(value());
  }

  return '';
}
