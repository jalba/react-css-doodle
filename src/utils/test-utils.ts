import { ReactTestRendererJSON } from 'react-test-renderer';
import { BasicTypes } from '../lib/types';
import { screen, Nullish, MatcherFunction } from '@testing-library/react';

type Renderer = ReactTestRendererJSON | ReactTestRendererJSON[] | null;

function isRenderer(arg: Renderer): arg is ReactTestRendererJSON {
  return (<ReactTestRendererJSON>arg).children !== undefined;
}

export const findInChildren = (json: Renderer, value: BasicTypes): boolean => {
  let found;

  if (isRenderer(json)) {
    found = json.children?.find(child =>
      child.toString().includes(value.toString())
    );
  }

  return Boolean(found);
};

// Idea from https://polvara.me/posts/five-things-you-didnt-know-about-testing-library

export const getByTextContent = (text: string): HTMLElement => {
  const { getByText } = screen;

  return getByText((_content, node) => {
    const hasText = (element: Nullish<Element>) =>
      element?.textContent?.includes(text);
    const nodeHasText = hasText(node);
    const nodeChildren = node ? node.children : [];
    const childrenDontHaveText = Array.from(nodeChildren).every(
      child => !hasText(child)
    );

    return Boolean(nodeHasText && childrenDontHaveText);
  });
};
