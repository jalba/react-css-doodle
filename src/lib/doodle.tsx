import React from 'react';

import DoodleComponent from '../components/doodle-component';

import { Vars } from './types';
import { valueToString } from '../utils';

export default function doodle(
  stringArray: TemplateStringsArray,
  ...vars: Vars[]
): () => JSX.Element {
  const computedRules: string = stringArray
    .map((str, idx) => `${str}${vars[idx] ? valueToString(vars[idx]) : ''}`)
    .join('');

  const Doodle = () => <DoodleComponent rules={computedRules} />;

  return Doodle;
}
