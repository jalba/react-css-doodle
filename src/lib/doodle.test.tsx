import React from 'react';
import renderer from 'react-test-renderer';

import { findInChildren } from '../utils';

import doodle from './doodle';

describe('the doodle function', () => {
  it('should render a doodle correctly', () => {
    const TwoByTwoGrid = doodle`
      :doodle {
        @grid: 2 / 200px;
        grid-gap: 1px;
      }
    `;
    const doodleTree = renderer.create(<TwoByTwoGrid />).toJSON();
    expect(doodleTree).toMatchSnapshot();
  });

  it('should render variables passed to the template string correctly', () => {
    const dimension = 4;
    const GridNumber = doodle`
      :doodle {
        @grid: ${dimension} / 200px;
        grid-gap: 1px;
      }
    `;
    const str = 'pink';
    const GridString = doodle`
      :doodle {
        @grid: ${dimension} / 200px;
        grid-gap: 1px;
      }
      background: ${str};
    `;
    const template = `background: @pick(red, ${str});`;
    const GridTemplate = doodle`
      :doodle {
        @grid: ${dimension} / 200px;
        grid-gap: 1px;
      }
      ${template}
    `;
    const GridFunction = doodle`
      :doodle {
        @grid: ${() => dimension * dimension} / 200px;
        grid-gap: 1px;
      }
    `;
    const numberJson = renderer.create(<GridNumber />).toJSON();
    const stringJson = renderer.create(<GridString />).toJSON();
    const templateJson = renderer.create(<GridTemplate />).toJSON();
    const functionJson = renderer.create(<GridFunction />).toJSON();
    const strNumberTest = `@grid: ${dimension} / 200px;`;
    const strStringTest = `background: ${str};`;
    const strFunctionTest = `@grid: ${dimension * dimension} / 200px;`;

    expect(findInChildren(numberJson, strNumberTest)).toEqual(true);
    expect(findInChildren(stringJson, strStringTest)).toEqual(true);
    expect(findInChildren(templateJson, template)).toEqual(true);
    expect(findInChildren(functionJson, strFunctionTest)).toEqual(true);
  });
});
