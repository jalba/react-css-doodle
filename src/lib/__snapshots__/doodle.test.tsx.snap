// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the doodle function should render a doodle correctly 1`] = `
<css-doodle>
  
      :doodle {
        @grid: 2 / 200px;
        grid-gap: 1px;
      }
    
</css-doodle>
`;
